aws/logs_monitoring/lambda_function.py

```diff
@@ -55,98 +74,230 @@
+            raise Exception("could not scrub the payload")
+        except Exception as e:
+            # most likely a network error
+            raise RetriableException()
+        if resp.status_code >= 500:
+            # server error
+            raise RetriableException()
+        elif resp.status_code >= 400:
+            # client error
+            raise Exception(
+                "client error, status: {}, reason {}".format(
+                    resp.status_code, resp.reason
+                )
+            )
+        else:
+            # success
+            return
 
     def __enter__(self):
-        self._sock = self._connect()
+        self._connect()
         return self
 
     def __exit__(self, ex_type, ex_value, traceback):
-        if self._sock:
-            self._sock.close()
-        if ex_type is not None:
-            print("DatadogConnection exit: ", ex_type, ex_value, traceback)
+        self._close()
+
+
+class DatadogBatcher(object):
+    def __init__(self, max_log_size_bytes, max_size_bytes, max_size_count):
+        self._max_log_size_bytes = max_log_size_bytes
+        self._max_size_bytes = max_size_bytes
+        self._max_size_count = max_size_count
+
+    def _sizeof_bytes(self, log):
+        return len(json.dumps(log).encode("UTF-8"))
+
+    def batch(self, logs):
+        """
+        Returns an array of batches.
+        Each batch contains at most max_size_count logs and
+        is not strictly greater than max_size_bytes.
+        All logs strictly greater than max_log_size_bytes are dropped.
+        """
+        batches = []
+        batch = []
+        size_bytes = 0
+        size_count = 0
+        for log in logs:
+            log_size_bytes = self._sizeof_bytes(log)
+            if size_count > 0 and (
+                size_count >= self._max_size_count
+                or size_bytes + log_size_bytes > self._max_size_bytes
+            ):
+                batches.append(batch)
+                batch = []
+                size_bytes = 0
+                size_count = 0
+            if log_size_bytes <= self._max_log_size_bytes:
+                # all logs exceeding max_log_size_bytes are dropped here
+                batch.append(log)
+                size_bytes += log_size_bytes
+                size_count += 1
+        if size_count > 0:
+            batches.append(batch)
+        return batches
+
+
+class ScrubbingRule(object):
+    def __init__(self, regex, placeholder):
+        self.regex = regex
+        self.placeholder = placeholder
+
+
+class DatadogScrubber(object):
+    def __init__(self, configs):
+        rules = []
+        for config in configs:
+            try:
+                if os.environ.get(config.name, False):
+                    rules.append(
+                        ScrubbingRule(
+                            re.compile(config.pattern, re.I), config.placeholder
+                        )
+                    )
+            except Exception:
+                pass
```

Comment:
log the exception here

Reply:
raise an exception instead, we should not forward the logs if we can't build the rule, the user should fix the pattern instead, thoughts ?